### PR TITLE
Add, Refactor, and Expand type indicator-base ShellItems

### PR DESCRIPTION
**Pull request by klaki892**
Created Skeleton of new ShellItemList for ShellItem distribution
Refactors ShellItem classes and Expansion Block implementations
Expanded ShellItem Definitions to obtain more properties
Adds tests for known ShellItem examples
full enumeration of ExpansionBlock properties still pending
Adds IntelliJ user config to .gitignore

Provides Significant progress to #51

### CODE REVIEW

File: WPF/SeeShells/SeeShells/ShellParser/ShellItems/ShellItemList.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        public ShellItemList(byte[] buf, int offset) : base(buf, offset) { }
+
+        protected  IShellItem GetItem(int off)
+        {
+            //item to return
+            IShellItem item = null;
+            //the shell item type which can be identified by 2 bytes (0xXX)
+            int type = unpack_byte(off + 2);
+
+            //TODO: Aleks - Dynamically call upon correct ShellItem0xXX to be instantiated
```
Comment (klaki892) [inline WPF/SeeShells/SeeShells/ShellParser/ShellItems/ShellItemList.cs]: Pointing this out for @AlekStoyanov 

We messed up in our original design. It turns out that the `IRegistryReader.getValue() Will not return a ShellItem ready to be decoded but instead a [Shell Item List].(https://github.com/libyal/libfwsi/blob/master/documentation/Windows%20Shell%20Item%20format.asciidoc#2-shell-item-list)

Basically you'll need to create a new `ShellItemList` for each registry value, and implement this method so you can ultimately return the list of `IShellItem`s to the greater good.

File: WPF/SeeShells/SeeShells/ShellParser/ShellItems/ShellItemList.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+            return item;
+        }
```
Comment (klaki892) [inline WPF/SeeShells/SeeShells/ShellParser/ShellItems/ShellItemList.cs]: Pointing this out for @AlekStoyanov 

We messed up in our original design. It turns out that the `IRegistryReader.getValue()` Will not return a ShellBag Item ready to be decoded but instead a [Shell Item List](https://github.com/libyal/libfwsi/blob/master/documentation/Windows%20Shell%20Item%20format.asciidoc#2-shell-item-list) object.

Basically you'll need to create a new `ShellItemList` for each registry value, and implement this method so you can ultimately return the list of `IShellItem`s to the greater good.

File: WPF/SeeShells/SeeShells/ShellParser/ShellItems/Block.cs [hunk 3]
```diff
@@ -2,21 +2,21 @@
...
-        public Block(byte[] buf, int offset, object parent)
+        protected byte[] buf { get; set; }
+        public int offset { get; protected set; }
+        protected object parent { get; set; }
+        protected Block(byte[] buf, int offset)
         {
             this.buf = buf;
             this.offset = offset;
-            this.parent = parent;
+            parent = parent;
```
Comment (szczura) [inline WPF/SeeShells/SeeShells/ShellParser/ShellItems/Block.cs]: isn't parent not changing here?

Comment (klaki892) [reply WPF/SeeShells/SeeShells/ShellParser/ShellItems/Block.cs]: Yup looks like a useless self-assignment to me. 
Happened while taking parent out of the constructor.
fixed.

File: WPF/SeeShells/SeeShells/ShellParser/ShellItems/ShellItemList.cs (hunk 2 above)
Comment (AlekStoyanov) [reply WPF/SeeShells/SeeShells/ShellParser/ShellItems/ShellItemList.cs]: By `IRegistryReader.getValue()` do you mean `IRegistryReader.GetRegistryKeys()` ?

So, my understanding is, the `ShellBagParser` `GetShellItems` method should create an instance of `ShellItemList` passing it the raw byte array from the `IRegistryReader` and call `Items` method in order to obtain a list of `IShellItem`. Meanwhile all the Activator reflection magic should happen in the `GetItem` method in `ShellItemList`

Comment (klaki892) [reply WPF/SeeShells/SeeShells/ShellParser/ShellItems/ShellItemList.cs]: Yes, everything you've said is correct.

**Pull request merged**